### PR TITLE
libblkid: Check LUKS OPAL before RAID

### PR DESCRIPTION
**Pull request by oldium**
When LUKS OPAL HW encryption is enabled and the partition is locked, any read from the locked area ends in reading error. The blkid tool is used in udev rules (at least in Debian), which is executed on every close of the block device, so errors from blkid can flood the logs. Fix this by checking LUKS first.

I do not know if it is correct to check LUKS before RAID, but it is crucial to not touch LUKS area before LUKS it is unlocked.

Not checking LUKS before RAID leads to the following errors when OPAL area is not yet unlocked:

```
kernel: ata8.00: exception Emask 0x0 SAct 0x0 SErr 0x0 action 0x0
kernel: ata8.00: irq_stat 0x40000001
kernel: ata8.00: failed command: READ DMA EXT
kernel: ata8.00: cmd 25/00:08:80:af:c0/00:00:d1:01:00/e0 tag 23 dma 4096 in
                  res 51/04:08:80:af:c0/00:00:d1:01:00/e0 Emask 0x1 (device error)
kernel: ata8.00: status: { DRDY ERR }
kernel: ata8.00: error: { ABRT }
kernel: I/O error, dev sdd, sector 7814033280 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 2
kernel: Buffer I/O error on dev sdd1, logical block 976753904, async page read
```

### CODE REVIEW

**Comment by oldium** [comment]
It would be possible to check for RAID before LUKS, but as far as I can tell only under the following conditions:

* No LUKS2 header is found or is invalid
* LUKS2 header is found and valid and does not contain OPAL encryption

**Comment by t-8ch** [comment]
FYI: This change was done before and reverted, see #2373 and #2061.
In the past the problem to be solved was only a misprobe, not the kernel message.

Instead of running the normal luks prober first, maybe we need a new opal_luks prober that only runs `if blkdid_probe_is_opal_locked()`.

**Comment by oldium** [comment]
Ha! Exactly what I was looking for. 😁

I just made a test with RAID 1 with mdadm metadata version 1.0 and LUKS clearly cannot precede, the RAID is not correctly detected first. I will try the code from #2889.

**Comment by mbroz** [comment]
@karelzak, please do not merge this (and #2889) until @oniko or I approve it. The whole detection is tricky, and we already broke storage priorities several times in the past. No need to repeat it again....

I think it is the correct way to fix this, but it needs at least detection of the LUKS2 subsystem label (we set it to HW-OPAL).

And I would definitely prefer if kernel rejects these reads properly (if you setup Opal or any other TCG SSC like Pyrite manually, it will end with the same problem. But this requires kernel patching and perhaps some discussion.)

**Comment by oldium** [comment]
The latest version checks for LUKS2 HW OPAL encryption header value and OPAL disk locking status and completes the probe only when the disk is locked. I think checking LUKS2+OPAL before RAID is safe, because the area of LUKS2 is completely unreadable unless you unlock it first. So there cannot be any RAID superblock anywhere inside LUKS area, because it is simply unreadable until you unlock it. In other words, you have to unlock LUKS2 to be able to detect any other superblocks.

**Comment by oldium** [comment]
One thing - if the secondary_offsets is used (non-zero header offset), there might be some superblock preceding LUKS. I will update the patch to specifically check for offset 0.

**Comment by oldium** [comment]
Done. Ready for review.

**Comment by oldium** [comment]
I would also like to see better handling of the locked area inside the kernel but seeing that there are unfixed nonsensical errors generated from the SATA layer (https://lore.kernel.org/all/20230623181908.2032764-1-lorenz@brun.one/ - yes, I see them exactly as described), I am a bit skeptical about the duration of the implementation. So getting one or two tools working with OPAL before the disk is unlocked during the boot sequence seems like it has a better chance to succeed. I just need to unlock the disk as soon as possible (I do it inside initrd environment) and every other read will simply succeed.

File: libblkid/src/superblocks/luks.c
Comment (mbroz) [inline libblkid/src/superblocks/luks.c:37]: This must be a C string, you must chcek also trailing \0. (With memcmp just add it here.)

File: libblkid/src/superblocks/luks.c
Comment (mbroz) [inline libblkid/src/superblocks/luks.c:178]: Why do you scan for the secondary header and throw it away if found? Just scan the primary one, then.

If we want this to be the first check in the blkid chain, it must be absolutely minimal and fast. I would probably simplify checking for only the first superblock on offset 0, only version 2, validity offset check and subsystem string. Perhaps a small function (duplication of existing code) is more readable here than the rewrite above.

File: libblkid/src/superblocks/luks.c
Comment (mbroz) [inline libblkid/src/superblocks/luks.c:186]: memcmp argument is void, cast is not needed. (and it must check for trailing \0 as mentioned above).

**Review by mbroz — COMMENTED** [review]
Some comments, just from POV of LUKS maintainer. As mentioned, it would be more readable with slight code duplication than tryuing to reuse existing secondary header magic.

File: libblkid/src/superblocks/luks.c
Comment (oldium) [reply libblkid/src/superblocks/luks.c:186]: Ah, copy-paste from luks_attributes.

File: libblkid/src/superblocks/luks.c
Comment (oldium) [reply libblkid/src/superblocks/luks.c:37]: This acts as a C-string. `printf("%u\n", (unsigned) sizeof(LUKS2_HW_OPAL_SUBSYSTEM));` outputs `8`, which is correct.

File: libblkid/src/superblocks/luks.c
Comment (oldium) [reply libblkid/src/superblocks/luks.c:178]: Thanks, looks reasonable.

**Comment by oldium** [comment]
Done. Ready for re-review 😊

**Comment by oldium** [comment]
And yes, I like it more 😁. Thanks for review!

**Comment by mbroz** [comment]
Yes, that's how I think it should be done (at least the LUKS part).  So kernel log is now clean?

(I think we need to discuss this later this week with @oniko, I'll ping Karel then.)

Thanks.

**Comment by oldium** [comment]
> Yes, that's how I think it should be done (at least the LUKS part). So kernel log is now clean?

I am going to make myself a patched deb package in order to test this during system boot tomorrow. Otherwise it runs clean.

**Review by bluca — APPROVED** [review]

**Comment by oldium** [comment]
The logs are clean now at least within my Debian (mostly bookworm) - I tried to compile a patched version from experimental tree on it (it did not go smooth, but with some packages from bookworm-backports it worked). The best of course would be to have this handled inside the kernel, which I doubt will happen any time soon.

**Comment by bluca** [comment]
Technically you can use it without cryptsetup, and instead using the yucky tsg tools that reimplement all the nvme commands in userspace, but then most likely that would be a full disk locking with the fake mbr thingy, rather than partition-based, so I'm not sure in which use cases it would be useful to support it here

**Comment by oldium** [comment]
@karelzak Would it be possible to get this into an official release (possibly a new 2.41), please? Thanks.

**Comment by karelzak** [comment]
> @karelzak Would it be possible to get this into an official release (possibly a new 2.41), please? Thanks.

It's already merged, so yes, it will be in v2.41 (hopefully in February).

**Comment by oldium** [comment]
Perfect, thanks.

**Pull request merged**
